### ISSUE TITLE
ingressController with nodePort Connection reset on 1.15

### ISSUE THREAD

**Issue by AntoineFr** [cilium/cilium#30944]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

When upgrading the Helm chart to 1.15.1 the nodePort I expose for Ingress can't be reached anymore on the local node. Rolling back to version 1.14.7 works.

1.15.1 ❌
```
$ curl localhost:30080 -v
*   Trying 127.0.0.1:30080...
* Connected to localhost (127.0.0.1) port 30080 (#0)
> GET / HTTP/1.1
> Host: localhost:30080
> User-Agent: curl/7.88.1
> Accept: */*
>
* Recv failure: Connection reset by peer
* Closing connection 0
curl: (56) Recv failure: Connection reset by peer
```

1.14.7 ✔️
```
$ curl localhost:30080 -v
*   Trying 127.0.0.1:30080...
* Connected to localhost (127.0.0.1) port 30080 (#0)
> GET / HTTP/1.1
> Host: localhost:30080
> User-Agent: curl/7.88.1
> Accept: */*
>
< HTTP/1.1 404 Not Found
< date: Sat, 24 Feb 2024 17:20:09 GMT
< server: envoy
< content-length: 0
<
* Connection #0 to host localhost left intact
```

I used these commands for the upgrade then rollback:
```
helm upgrade --namespace kube-system cilium cilium/cilium --version 1.15.1 -f values.yaml
helm upgrade --namespace kube-system cilium cilium/cilium --version 1.14.7 -f values.yaml
```

With this file `values.yaml`: 
```yaml
ingressController:
  default: true
  enabled: true
  loadbalancerMode: shared
  service:
    insecureNodePort: 30080
    secureNodePort: 30443
    type: NodePort
k8sServiceHost: 192.168.1.50
k8sServicePort: 6443
kubeProxyReplacement: true
operator:
  replicas: 1
```

### Cilium Version

1.15.1

### Kernel Version

Linux raspberrypi5 6.1.0-rpi7-rpi-2712 #1 SMP PREEMPT Debian 1:6.1.63-1+rpt1 (2023-11-24) aarch64 GNU/Linux

### Kubernetes Version

Client Version: v1.28.6
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.6

### Regression

1.14.7

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by chaunceyjiang** [comment]
> helm upgrade --namespace kube-system cilium cilium/cilium --version 1.15.1 -f values.yaml

```
root@node1:~# helm list -n kube-system
NAME      	NAMESPACE  	REVISION	UPDATED                                	STATUS  	CHART           	APP VERSION
cilium    	kube-system	2       	2024-02-26 14:16:39.181966045 +0800 CST	deployed	cilium-1.15.1   	1.15.
```

I can't reproduce this issue. Can you provide more additional information?

```
root@node1:~# curl localhost:30080 -v
*   Trying 127.0.0.1:30080...
* Connected to localhost (127.0.0.1) port 30080 (#0)
> GET / HTTP/1.1
> Host: localhost:30080
> User-Agent: curl/7.81.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 503 Service Unavailable
< content-length: 152
< content-type: text/plain
< date: Mon, 26 Feb 2024 06:21:31 GMT
< server: envoy
<
* Connection #0 to host localhost left intact
```

**Comment by AntoineFr** [comment]
Hello @ldelossa , here are the dumps before and after the upgrade.

- `cilium-sysdump-20240227-190905.zip` is before (1.14.7)
- `cilium-sysdump-20240227-191235.zip` is after (1.15.1)

I am trying to investigate a bit more to find a cause but I can't find anything in logs for example.

**Comment by tskinner-oppfi** [comment]
Here are some repro steps via kind:

```
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
- role: worker
  extraPortMappings:
  - containerPort: 30000
    hostPort: 8080
networking:
  disableDefaultCNI: true
```

```
kind create cluster --cluster-config <abovefile>
```

Install helm

```
helm upgrade cilium cilium/cilium --version 1.15.1 --install \
   --namespace kube-system \
   --set image.pullPolicy=IfNotPresent \
   --set ipam.mode=kubernetes \
   --set ingressController.enabled=true \
   --set ingressController.service.type=NodePort \
   --set ingressController.service.insecureNodePort=30000 \
   --set ingressController.loadbalancerMode=shared
```

Apply the following:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: httpbin
  name: httpbin
spec:
  replicas: 1
  selector:
    matchLabels:
      app: httpbin
  template:
    metadata:
      labels:
        app: httpbin
    spec:
      containers:
      - env:
        - name: MAX_DURATION
          value: 300s
        image: mccutchen/go-httpbin
        name: go-httpbin
        ports:
        - containerPort: 8080
          name: http
          protocol: TCP
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: httpbin
  name: httpbin
spec:
  ports:
  - name: 8080-8080
    port: 8080
    protocol: TCP
    targetPort: 8080
  selector:
    app: httpbin
  type: ClusterIP
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: httpbin
  namespace: default
spec:
  ingressClassName: cilium
  rules:
  - host: httpbin.test.com
    http:
      paths:
      - backend:
          service:
            name: httpbin
            port:
              number: 8080
        path: /
        pathType: Prefix

```

Run curl and the connection gets closed.

```
curl -v -H 'Host:httpbin.test.com' 'http://localhost:8080/drip?duration=20&numbytes=20&&delay=0'
```

In my more prod-like setup this happens with tls termination and it is more consistently cut off at 15 seconds. You can play with the httpbin query parameters but same results every time if the response takes longer than 15 seconds. [http bin docs](https://httpbin.org/#/Dynamic_data/get_drip)

**Comment by AntoineFr** [comment]
Hello, I fixed the issue by upgrading to 1.16 and using the new options provided by this PR : https://github.com/cilium/cilium/pull/30840
I'm therefore closing this issue.
